packages/flutter/lib/src/material/focusable.dart

```diff
@@ -0,0 +1,297 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/painting.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+import 'theme.dart';
+
+typedef FocusableOnKeyCallback = bool Function(FocusableNode node, RawKeyEvent event);
+
+/// A widget that manages a [FocusableNode] and allows specification of the
+/// focus order.
+///
+/// The node can be either a scope node or a leaf node. Scope nodes provide a
+/// scope for their children, using the focus traversal policy defined by the
+/// [DefaultFocusTraversal] widget above them to traverse their children.
+///
+/// It provides [onFocusChange] as a way to be notified when the focus is given
+/// to or removed from this widget, and allows specification of a [focusedDecoration]
+/// to be shown when its [child] has focus.
+///
+/// The [onKey] argument allows specification of a key even handler that should
+/// be invoked when this node or one of its children has focus.
+class Focusable extends StatefulWidget {
```

Comment:
would it make sense to have this in the `widgets` layer rather than `material`?

Reply:
It would, and I started with it there, but for the focus highlighting, it needs Themes, so there's a layer violation if I put it there, so it had to be here.  I could make a "MaterialFocusable" that uses themes, and have the bulk of Focusable live in widgets, but that didn't seem useful.  Another thought I had was to just not do highlighting at all, and have another widget in material that listens to the focusable node above it and does highlighting.